### ISSUE TITLE
Redis Cluster Support?

### ISSUE THREAD

**Issue by BigBorg** [ionelmc/python-redis-lock#59]
Does this package support Redis Cluster? I find that to make atomic operation work with multiple keys, {} brackets would be needed, which makes "lock:resource" as "lock:{resource}" and "lock-signal:resource" as "lock-signal:{resource}".

The Redis Cluster document is as follows:

Redis Cluster supports multiple key operations as long as all the keys involved into a single command execution (or whole transaction, or Lua script execution) all belong to the same hash slot. The user can force multiple keys to be part of the same hash slot by using a concept called hash tags.

Hash tags are documented in the Redis Cluster specification, but the gist is that if there is a substring between {} brackets in a key, only what is inside the string is hashed, so for example this{foo}key and another{foo}key are guaranteed to be in the same hash slot, and can be used together in a command with multiple keys as arguments.

**Comment by ionelmc** [comment]
Perhaps you meant `{lock}resource` and ``{lock-signal}resource`?

**Comment by BigBorg** [comment]
No, only content inside {} bracket is used to generate hash slot. If you use  {lock}resource and {lock-signal}resource, they might end up in different slot since hash result of "lock" and "lock-signal" is different. But if you use lock:{resource} and lock-signal:{resource}, the two keys for the same resource will end up in the same slot. Also, since resource differs, this will spread out lock and lock-signal keys for different resources which is the purpose of Redis Cluster ( to shard content ).

**Comment by ionelmc** [comment]
Well I guess we could change it but then a new release would break everything for most people. A better way would be a way to allow users to override the formatting (so you could do your RC stuff in your lock subclass).

**Comment by ionelmc** [comment]
You can already do it right now, albeit a bit ugly:
```python

class RedisClusterLock(Lock):
  def __init__(self, *args, **kwargs):
    super(RedisClusterLock, self).__init__(*args, **kwargs)
    self._name = 'lock:{%s}' % name
    self._signal = 'lock-signal:{%s}' % name
```

**Comment by ionelmc** [comment]
@amielboim isn't that subclass working for you? I mean I could add a "cluster=False" option so ppl can opt-in to the new key names if you want something simpler. I can't simply change the lock names by default since it would make upgrading the library troublesome.
